### PR TITLE
Fix metatest.c with gcc-12 -Wuse-after-free

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Fix the build of `metatest.c` under `gcc-12 -Wall`: `-Wuse-after-free` detects the deliberate use-after-free that we want to occur at runtime (because we're testing the runtime detection by Asan).

Although we do run GCC 12 on the CI, we only run it in `all.sh test_gcc_latest`, and that uses the `full` configuration. In that configuration, `mbedtls_free` is a function pointer variable in `platform.o` so the compiler doesn't see what's going on. In the  default configuration, `mbedtls_free` is a macro that expands to `free`, so the compiler can see that there's a use-after-free.

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** not required
- [x] **backport** https://github.com/Mbed-TLS/mbedtls/pull/8557
- [x] **tests** manually (see above)

### CODE REVIEW

**Review by davidhorstmann-arm — APPROVED** [review]
LGTM, thanks!

Reproduced with GCC 13 locally and checked that ASan metatests still work, along with standard visual review.

File: programs/test/metatest.c [hunk 1]
```diff
@@ -46,6 +46,12 @@
  */
 volatile int false_but_the_compiler_does_not_know = 0;
 
+/* Hide calls to calloc/free from static checkers such as
+ * `gcc-12 -Wuse-after-free`, to avoid compile-time complaints about
+ * code where we do mean to cause a runtime error. */
+void * (* volatile calloc_but_the_compiler_does_not_know)(size_t, size_t) = mbedtls_calloc;
+void(*volatile free_but_the_compiler_does_not_know)(void *) = mbedtls_free;
```
Comment (paul-elliott-arm) [inline programs/test/metatest.c:53]: Surely one of these is going to annoy the code style check?

**Review by paul-elliott-arm — COMMENTED** [review]
LGTM, confirms fixing the problem, however suspect some code style issues.

File: programs/test/metatest.c (hunk 1 above)
Comment (paul-elliott-arm) [reply programs/test/metatest.c:53]: Apparently not, but can we standardise on one style?

Comment (gilles-peskine-arm) [reply programs/test/metatest.c:53]: [This feels inconsistent to me too](https://arm-ce.slack.com/archives/GTM3SM1K5/p1700570538230139), but I'm not going to expend energy to improve the presentation of function pointer declarations, which come up very rarely in our code base.

Comment (paul-elliott-arm) [reply programs/test/metatest.c:53]: Oh. Right. I misunderstood, uncrustify is actually **enforcing** this, not being lax about it. In which case nvm

**Review by paul-elliott-arm — APPROVED** [review]
LGTM

**Pull request merged**